docs/PLUGIN.md

```diff
@@ -1,19 +1,195 @@
+3. A command-line application host will look for plugin configuration in
+   environment variables or the appSettings file, then will use the configured
+   plugin names to load .NET assemblies from the runtime binary directory.
+
+### Backend Architecture
+
+The core Data Management Service business logic will forward data access and
+modification requests to a class implementing one of two interfaces, which are
+defined in `EdFi.DataManagementService.Core.External.Interface`:
+
+* `IDocumentStoreRepository` supports create, update, and delete functionality,
+  as well as reading by ID.
+* `IQueryHandler` supports read all or read by query requests.
+
+The "GET by ID" functionality is separated from the query handling to protect
+against eventual consistency confusion. In the case where the query handler is
+using a different data store than the document repository, there may be a small
+lag between writes to each system. A client application that has just written a
+document would reasonably expect that it is immediately available for retrieval,
+hence the retrieval by ID will pull directly from the data store used for the
+write operations.
+
+> [!TIP]
+> These two interfaces could be implemented by the same C# class referencing the
+> same backend datastore, for example if a given deployment chooses to use
+> PostgreSQL for all storage and retrieval instead of using a search database.
+
+```mermaid
+classDiagram
+    class IQueryHandler
+        <<interface>> IQueryHandler
+        IQueryHandler: QueryDocuments(IQueryRequest queryRequest)
+
+    class IDocumentStoreRepository
+        <<interface>> IDocumentStoreRepository
+        IDocumentStoreRepository: UpsertDocument(IUpsertRequest upsertRequest)
+        IDocumentStoreRepository: GetDocumentById(IGetRequest getRequest)
+        IDocumentStoreRepository: UpdateDocumentById(IUpdateRequest updateRequest)
+        IDocumentStoreRepository: DeleteDocumentById(IDeleteRequest deleteRequest)
+```
+
+To implement a new backend:
+
+1. Create a new C# project.
+2. Create a dependency on library `EdFi.DataManagementService.Core.External`.
+3. Implement one or both interfaces described above.
+4. Document required app settings
+
+> [!WARNING]
+> More detail needed on using the plugin architecture. More will is probably
+> necessary to support arbitrary configuration information. The plugin architecture
+> might require an initialization step that accepts a configuration provider
+> that the plugin can query to retrieve needed information.
+
+### Frontend Architecture
+
+In order to support cloud function implementations, it will be useful to treat
+the ASP.NET frontend as one of many possible frontend implementations. To do so
+might require separation of the console application from the frontend
+implementation.
+
+Each of the big-three cloud providers' serverless function technologies have
+.NET libraries that can extend ASP.NET applications. For AWS and Azure, the
```

Comment:
Like I say, we are very close to being able to cloud deploy without being a heavy-weight ASP.NET Core app. I'm confident we'll be able to do things just like Meadowlark, with a pure AWS Lambda frontend.

Reply:
I agree, I'm confident that it won't take much to get there.